### ISSUE TITLE
IsByRefLike keep alocating values on heap

### ISSUE THREAD

**Issue by Wkalmar** [dotnet/fsharp#10216]
While type is decorated with `IsByRefLike` attribute it keeps allocating values on heap.

**Repro steps**

1. Checkout code provided [here](https://github.com/Wkalmar/Orckestra/tree/fsharp-span)
2. In visual studio run Debug -> Performance profiler -> .NET object allocation tracking
3. Observe allocation output

**Expected behavior**

No values from `Orckestra.Fsharp` namespace are allocated.

**Actual behavior**

![allocations](https://user-images.githubusercontent.com/7817214/94995571-e5eb1c00-05a7-11eb-9375-7a48bb84c8a6.PNG)

**Related information**

I've tried to simplify repro code to not include all these C#-F# interactions, but there was no adequate tooling to check heap allocations for F# in Visual Studio.

* Operating system
* .NET standard 2.1, Windows 10
* Visual studio community 2018

**Bonus**
1. Uncomment benchmark code in main
```
    class Program
    {
        static void Main(string[] args)
        {
            BenchmarkRunner.Run<Test>();
            Console.ReadLine();

        }
    }
```
2. Run the application
3. Observe that F# version is way slower

I hope you're having a good day,
Bohdan

**Comment by cartermp** [comment]
I haven't profiled anything myself, but there are two things here:

1. `PermissionInfo` itself isn't marked as `IsByRefLike`
2. The array where `PermissionInfo` is being allocated in the code still needs to be allocated

I'll need some more information about what you're trying to do here though. By having an array literal of structs, you're still allocating. There's no getting around that. The question is if you're continuing to allocate things by processing that array. Is that what you are seeing?

**Comment by Wkalmar** [comment]
Thank you for your response, Phillip
> PermissionInfo itself isn't marked as IsByRefLike

This is a fair point. However, marking it as `IsByRefLike` gives me compiler error "A type instantiation involves a byref type. This is not permitted by the rules of Common IL." I guess it's due to the fact that `ref struct` can't be a generic argument

> By having an array literal of structs, you're still allocating. There's no getting around that.

You've just helped me realize that my C# version also keeps allocating despite the fact that it's a field of a `ref struct`. But it does it less since `Permissions` is `static`. 

Sorry for taking your time.

**Comment by Wkalmar** [comment]
However, I'm still struggling to wrap my head around it. I've changed my field to `static member private Permissions` and it is still allocated 378 times. While in my C# version this number is 3.
> I'll need some more information about what you're trying to do here though.

Business-wise I'm trying to convert linux permission string into digit. But the thing is that linux permission string is huge. In fact, what I'm trying to do is experimenting with `Span<T>` and trying to squeeze some performance of it. So what I've ended up doing is parsing a string into C# type `SymbolicPermission3` which does some validation and internally holds `ReadOnlySpan<char> _value`. 

Then I've decided to delegate some conversion work to F# module `Helper` that accepts `ReadOnlySpan` to check whether F# can keep up in terms of performance.

So I've run a benchmark comapring it with pure C# versions that use `Span` and don't use it and noticed that the code that uses F# helper is way slower and the number of allocations is way bigger. So I've investigated it a little bit and found these `PermissionInfo[]` allocations quite suspicious.
